torch/distributions/multivariate_normal.py

```diff
@@ -0,0 +1,155 @@
+
+    This function takes as input `bmat`, containing :math:`n \times n` matrices, and
+    `bvec`, containing length :math:`n` vectors.
+
+    Both `bmat` and `bvec` may have any number of leading dimensions, which correspond
+    to a batch shape. They are not necessarily assumed to have the same batch shape,
+    just ones which can be broadcasted.
+    """
+    n = bvec.size(-1)
+    batch_shape = _get_batch_shape(bmat, bvec)
+
+    # to conform with `torch.bmm` interface, both bmat and bvec should have `.dim() == 3`
+    bmat = bmat.expand(batch_shape + (n, n)).contiguous().view((-1, n, n))
+    bvec = bvec.unsqueeze(-1).expand(batch_shape + (n, 1)).contiguous().view((-1, n, 1))
+    return torch.bmm(bmat, bvec).squeeze(-1).view(batch_shape + (n,))
+
+
+def _batch_potrf_lower(bmat):
+    """
+    Applies a Cholesky decomposition to all matrices in a batch of arbitrary shape.
+    """
+    n = bmat.size(-1)
+    cholesky = torch.stack([C.potrf(upper=False) for C in bmat.unsqueeze(0).contiguous().view((-1, n, n))])
+    return cholesky.view(bmat.shape)
+
+
+def _batch_diag(bmat):
+    """
+    Returns the diagonals of a batch of square matrices.
+    """
+    return bmat.contiguous().view(bmat.shape[:-2] + (-1,))[..., ::bmat.size(-1) + 1]
+
+
+def _batch_mahalanobis(L, x):
+    r"""
+    Computes the squared Mahalanobis distance :math:`\mathbf{x}^\top\mathbf{M}^{-1}\mathbf{x}`
+    for a factored :math:`\mathbf{M} = \mathbf{L}\mathbf{L}^\top`.
+
+    Accepts batches for both L and x.
+    """
+    # TODO: use `torch.potrs` or similar once a backwards pass is implemented.
+    flat_L = L.unsqueeze(0).contiguous().view((-1,) + L.shape[-2:])
+    L_inv = torch.stack([torch.inverse(Li.t()) for Li in flat_L]).view(L.shape)
+    return (x.unsqueeze(-1) * L_inv).sum(-2).pow(2.0).sum(-1)
+
+
+class MultivariateNormal(Distribution):
+    r"""
+    Creates a multivariate normal (also called Gaussian) distribution
+    parameterized by a mean vector and a covariance matrix.
+
+    The multivariate normal distribution can be parameterized either
+    in terms of a positive definite covariance matrix :math:`\mathbf{\Sigma}`
+    or a lower-triangular matrix :math:`\mathbf{L}` with positive-valued
+    diagonal entries, such that
+    :math:`\mathbf{\Sigma} = \mathbf{L}\mathbf{L}^\top`. This triangular matrix
+    can be obtained via e.g. Cholesky decomposition of the covariance.
+
+    Example:
+
+        >>> m = MultivariateNormal(torch.zeros(2), torch.eye(2))
+        >>> m.sample()  # normally distributed with mean=`[0,0]` and covariance_matrix=`I`
+        -0.2102
+        -0.5429
+        [torch.FloatTensor of size 2]
+
+    Args:
+        loc (Tensor or Variable): mean of the distribution
+        covariance_matrix (Tensor or Variable): positive-definite covariance matrix
+        scale_tril (Tensor or Variable): lower-triangular factor of covariance, with positive-valued diagonal
+
```

Comment:
Considering the recent merging of Tensor and Variable classes, could you please update this?

Reply:
Could you clarify how exactly it should be updated? I'm a bit confused whether the new combined classes should be referred to as `Tensor` or as `Variable`. I tried looking for the updated docs, but I don't believe they've been merged into master yet.

Should we also be updating all example and test code to avoid the `Variable` wrapper?

Reply:
Variable will soon disappear and be replaced with Tensor